### PR TITLE
Recommender Protocol and Related Signature Changes

### PR DESCRIPTION
**Pull request by Scienfitz**
As discussed we intend to unify `Strategy` and `Recommender`
This is the first of three PRs intending to do so:

1. This PR
   - Introduce the `RecommenderProtocol`
   - Change all relevant signatures so both `Strategy` and `Recommender` are allowed under this protocol
   - This also required refactoring in the `simulation` module, `Campaign.recommend` and several examples
   - Also made `select_recommender` private as its an internal concept
   - Also performed two minor but long standing refactorings (see Azure work item 15200). @AdrianSosic @AVHopp It might be easier to assess these last two commits separately from the rest:
     - Make `_recommend_continuous` and `_recommend_discrete` accept their respective subspaces instead of the entire `SearchSpace` (2nd last commit)
     - Changed the signature of `BayesianRecommender.recommend` so it is the same as for other recommenders by making the acquisition function a private `init=False` field (last commit)

Upcoming:

2. Turn `Strategies` into `Recommender` objects under a reasonable name like eg `CompositeRecommender`'s etc

3. Update docs regarding these changes

I've removed one existing deprecation as it was meddling with the `cattrs` structuring
I've not introduced any new deprecations. i) Its questionable whether we want to flood the codebase just for the ability to use the old strategy interface and ii) even if we wanted, its reasonable to do so after 2) and not 1)

### CODE REVIEW

**Review by AdrianSosic — COMMENTED** [review]
Hi @Scienfitz, thanks for the draft – and great that you nicely structured the commit, makes it easy to follow 👏🏼 . Before you continue, let us have a quick sync about the next steps, to make sure we area heading toward the same goal. In fact, now looking at the result, I notice a few problems ... (btw: this is what I meant when I said "let's write the protocol and then we'll see if it makes sense or not)

**Comment by Scienfitz** [comment]
@AdrianSosic ok what are the issues you're seeing?

**Comment by Scienfitz** [comment]
@AdrianSosic so I've reverted the hanges related to the sugnature and mvoed them to `Recommender`. This however requires some small workaround because `Strategy` does not have the `allow` attributes but they are referenced ind `simulation`. Im not sure if its worth fixing this here as `simulation` is under reconstruction anyway AND strictly speaking the `allow` attributes cannot be simply set by `Strategies` because their recommenders could have varying stuff. This would also not be solved in the override-approach discussed. Seems to me simply simulation should not couple to the `allow` attributes

Ive also removed the abstract `recommend` method form `Recommender`. 
- Ive nut put default values in the protoco... dont you think thats weird?
- its also strange to me that this is allowed, now `Recommender` derives from `RecommenderProtocol` but doesnt implement the protocol...

File: baybe/recommenders/base.py [hunk 1]
```diff
@@ -31,7 +36,8 @@ def _select_candidates_and_recommend(
 
     Args:
         searchspace: The search space.
```
Comment (AVHopp) [inline baybe/recommenders/base.py:38]: Need to adjust the docstring here since this is now no longer a search space but a subspace.

File: baybe/recommenders/base.py [hunk 2]
```diff
@@ -138,28 +147,29 @@ def recommend(  # noqa: D102
...
                 batch_quantity,
-                allow_repeated_recommendations,
-                allow_recommending_already_measured,
+                self.allow_repeated_recommendations,
+                self.allow_recommending_already_measured,
             )
         if searchspace.type == SearchSpaceType.CONTINUOUS:
             return self._recommend_continuous(
-                searchspace=searchspace, batch_quantity=batch_quantity
+                subspace_continuous=searchspace.continuous,
```
Comment (AVHopp) [inline baybe/recommenders/base.py:155]: Is there any specific reason why these changes are made in this PR? To me, this feels like something rather unrelated which should thus be done in a different PR.

File: baybe/recommenders/bayesian.py [hunk 3]
```diff
@@ -48,6 +53,11 @@ class BayesianRecommender(Recommender, ABC):
     ] = field(default="qEI")
     """The used acquisition function class."""
 
+    _acquisition_function: Optional[AcquisitionFunction] = field(
```
Comment (AVHopp) [inline baybe/recommenders/bayesian.py:56]: I think you already explained at one point, but could you remind me why this change is necessary?

**Review by AVHopp — COMMENTED** [review]
Just some minor comments so far. Are there more changes to be expected here in this PR? Do not know what exactly we agreed upon in our last meeting.

File: baybe/recommenders/bayesian.py (hunk 3 above)
Comment (Scienfitz) [reply baybe/recommenders/bayesian.py:56]: please see see Azure work item 15200
this unifies the recommender signatures which were different for bayesian and other recommenders

File: baybe/recommenders/base.py (hunk 2 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:155]: please see see Azure work item 15200
these are also recommender refactoring so why do yous ay they are unrelated?

Comment (AVHopp) [reply baybe/recommenders/base.py:155]: Well, this is a *different* change and refactoring and not related to unifying strategies and recommenders

Comment (Scienfitz) [reply baybe/recommenders/base.py:155]: ok ive mentioned that in the PR description the 2 other refactoings were put into 1 commit each, if you insist they could be cherrypicked and put into another PR

Comment (AVHopp) [reply baybe/recommenders/base.py:155]: I'd personally prefer to have this as a separate PR, but I am also fine with keeping it here. What is @AdrianSosic opinion here?

File: baybe/recommenders/base.py [hunk 4]
```diff
@@ -81,23 +88,15 @@ def _select_candidates_and_recommend(
...
-    @abstractmethod
     def recommend(
         self,
         searchspace: SearchSpace,
-        batch_quantity: int = 1,
-        train_x: Optional[pd.DataFrame] = None,
-        train_y: Optional[pd.DataFrame] = None,
-        allow_repeated_recommendations: bool = False,
-        allow_recommending_already_measured: bool = True,
+        batch_quantity: int,
```
Comment (AdrianSosic) [inline baybe/recommenders/base.py:97]: Absolutely agree, having default values here wouldn't make sense. What do you suggest? Technically, we don't need to define them on the abstract level (protocol / base class) since all of them recommenders will have their specific values in their signatures anyway. But I'm open for ideas.

And BTW: Since `Recommender` is an ABC, I don't think it's crazy that it does not implement the protocol, but I also haven't seen any good practices for that. Have you done some quick search already?

**Review by AdrianSosic — APPROVED** [review]
Hi @Scienfitz, thanks for the PR. I am really happy that we are slowly solving my two big headaches in the framework (recommender/strategy architecture and, at some point, scaling 🙃 ). 

In my eyes, the PR and what we have agreed on makes perfectly sense now. But I hope I have not overshadowed your views with my idea. So let's take a last moment before merge and really reassure that this is going in the right direction. All I can say at the moment is:
* I like the way the new interface looks like. This exactly what I would expect a recommender to look like.
* I like the fact that both high- and low-level recommendation functionality share this interface.

Of course that doesn't mean we are overlooking certain aspects in the design.
So if you have any last words, now is the moment to speak up 😄 

Apart from that, below my (very short) review. Note that I also submitted a few changes. Have a look
and let me know if you spot problems. I see some more changes coming, but they can wait after this refactoring is complete.

File: baybe/recommenders/base.py (hunk 2 above)
Comment (AdrianSosic) [reply baybe/recommenders/base.py:156]: In the general case, I'd be with @AVHopp here, but since it's already done and I've also put more changes on top during my review, I think it would now be more dangerous to revert than just keeping the commits in this PR. So let's be pragmatic here, save some effort, and eat cookies instead 🍪

File: baybe/recommenders/base.py (hunk 4 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:97]: yes I would just leave defaults out of the protocol

Haven seen this particular method mentioned, but I found an article that examined Protocols and ABCs as two different ways of achieving roughly similar things. So in one sense it might be weird that we use both, in another I didnt really see big problems or unreadable code so I guess its fine

File: baybe/strategies/base.py [hunk 5]
```diff
@@ -62,4 +62,4 @@ def allow_recommending_already_measured(self):
 
 # Register (un-)structure hooks
 converter.register_unstructure_hook(Strategy, unstructure_base)
-converter.register_structure_hook(Strategy, structure_strategy)
+converter.register_structure_hook(Strategy, structure_recommender_protocol)
```
Comment (Scienfitz) [inline baybe/strategies/base.py:95]: is this necessary? Strategy already has a structure hook (via `RecommenderProtocol` inheritance) which is also the same function?

File: baybe/simulation.py [hunk 6]
```diff
@@ -446,10 +447,11 @@ def simulate_experiment(
           The individual measurements obtained for the respective target and iteration
     """
     # TODO: Due to the "..." operator, sphinx does not render this properly. Might
-    # want to investigate in the future.
-    # TODO: In the markdown variant, bullet points with more levels are not rendered
-    # properly. We thus might want to refactor this when using html based documentation
-    # Validate the lookup mechanism
+    #   want to investigate in the future.
+    # TODO: Use a `will_terminate` campaign property to decide if the campaign will
```
Comment (Scienfitz) [inline baybe/simulation.py:451]: question: is this really necessary? what would you use it for? The solution below of stopping the simulation a few iterations before 100% of points are measures is totally fine. Is there any reason why all simulations MUST go to 100%  (if `n_doe_iterations` was `None`)?

File: baybe/recommenders/base.py (hunk 2 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:156]: then lets just finish this PR ill resolve

File: CHANGELOG.md [hunk 7]
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Copy button for code blocks in documentation
 - `mypy` for campaign, constraints and telemetry
+- `RecommenderProtocol` as common interface for `Strategy` and `Recommender`
```
[the remark this answers is not in the document]
Comment (Scienfitz) [reply CHANGELOG.md:11]: seems unnecessary to me, the main branch is in a permanently unfinished state between releases (and due to de deprecations theres no hard break)

[the remark this answers is not in the document]
Comment (AdrianSosic) [reply CHANGELOG.md:11]: Ok, then let’s go ahead

File: baybe/recommenders/bayesian.py [hunk 8]
```diff
@@ -72,23 +83,22 @@ def _get_acquisition_function_cls(
 
     def setup_acquisition_function(
         self, searchspace: SearchSpace, train_x: pd.DataFrame, train_y: pd.DataFrame
-    ) -> AcquisitionFunction:
+    ) -> None:
         """Create the current acquisition function from provided training data.
 
+        The acquisition function is stored in the private attribute
+        `_acquisition_function`.
```
Comment (AVHopp) [inline baybe/recommenders/bayesian.py:90]: Use double backticks

File: baybe/simulation.py [hunk 9]
```diff
@@ -504,23 +495,15 @@ def simulate_experiment(
...
-            assert not allow_measured
-
-            measured, _ = campaign.searchspace.discrete.get_candidates(
-                allow_repeated_recommendations=allow_repeated,
-                allow_recommending_already_measured=allow_measured,
+            # TODO: There can be still N < batch_quantity points left in the search
+            #   space. Once the recommender/strategy refactoring is completed,
+            #   find an elegant way to return those.
+            warnings.warn(
+                "The simulation of the campaign ended because there were no more "
```
Comment (AVHopp) [inline baybe/simulation.py:502]: Change to "..because not sufficiently many points were left for recommendation" - this makes it (in my opinion) clearer for the case of `batch_size>1`

File: baybe/recommenders/base.py [hunk 10]
```diff
@@ -106,16 +105,29 @@ def recommend(
...
-                search spaces.
 
         Returns:
             A DataFrame containing the recommendations as individual rows.
         """
+        ...
+
+
+@define
+class Recommender(ABC, RecommenderProtocol):
```
Comment (AVHopp) [inline baybe/recommenders/base.py:116]: According to the intended usage of `Protocol`, it should not be necessary to use inheritance here. This should be double-checked!

File: baybe/strategies/base.py [hunk 11]
```diff
@@ -3,26 +3,48 @@
...
+from baybe.recommenders.base import Recommender, RecommenderProtocol
+from baybe.recommenders.deprecation import structure_recommender_protocol
 from baybe.searchspace import SearchSpace
 from baybe.serialization import SerialMixin, converter, unstructure_base
-from baybe.strategies.deprecation import structure_strategy
 
 
 @define
-class Strategy(SerialMixin, ABC):
+class Strategy(SerialMixin, RecommenderProtocol, ABC):
```
Comment (AVHopp) [inline baybe/strategies/base.py:18]: According to the intended usage of `Protocol`, it should not be necessary to use inheritance here. This might however be different compared to `Recommender` since this class lives in another file. This should be double-checked!

File: baybe/strategies/base.py [hunk 12]
```diff
@@ -3,26 +3,48 @@
...
 
-    allow_recommending_already_measured: bool = field(default=False, kw_only=True)
-    """Allow to output recommendations that were measured previously. This only has an
-    influence in discrete search spaces."""
+    allow_recommending_already_measured: bool = field(default=None, kw_only=True)
+    """Deprecated! The flag has become an attribute of
+    :class:`baybe.recommenders.base.Recommender`."""
+
+    @allow_repeated_recommendations.validator
+    def validate_allow_repeated_recommendations(self, _, value):
```
Comment (AVHopp) [inline baybe/strategies/base.py:30]: Should this (as well as the other corresponding functions) be public? This will make them appear in the documentation, so we should be sure that we want this.

File: baybe/strategies/base.py [hunk 13]
```diff
@@ -3,26 +3,48 @@
...
+        """Raise a DeprecationError if the flag is used."""
```
Comment (AVHopp) [inline baybe/strategies/base.py:31]: Put DeprecationError in backticks or make a link out of it (also at other places).

File: baybe/recommenders/base.py (hunk 10 above)
Comment (AVHopp) [reply baybe/recommenders/base.py:116]: Potentially related: The init function now has a really weird signature, both in the table and in the overview:
image
image
We should just check that this is fine for us resp. that we might want to remove the function from e.g. the table.

**Review by AVHopp — APPROVED** [review]
There are some things that need to be verified and some minor things that should be changed. However, I believe that this is hopefully not too much work and too much of an issue. To avoid blocking, I hence already approve. Ping me if stuff is unclear or needs more discussion.

File: baybe/strategies/base.py (hunk 13 above)
Comment (Scienfitz) [reply baybe/strategies/base.py:31]: done with backticks

File: baybe/strategies/base.py (hunk 11 above)
Comment (Scienfitz) [reply baybe/strategies/base.py:18]: keeping it, see comments in dev thread

File: baybe/simulation.py (hunk 9 above)
Comment (Scienfitz) [reply baybe/simulation.py:502]: fine with me

File: baybe/strategies/base.py (hunk 12 above)
Comment (Scienfitz) [reply baybe/strategies/base.py:30]: good catch, fixed

File: baybe/recommenders/base.py (hunk 10 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:116]: keeping inheritance as discussed in dev chat.

regarding the doc: i think this is partially because they are kw_only. I tried the look also without kw_only but that just removes the `*` and it mostly looks the same

to me, the init doesnt look that weird overall? Also, these should really be kw_only and changing back due to this minor thing imos doesnt make sense, hence ill resolve
